certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,408 @@
+        f.selector == popBack().selector
+    ) => emptyBefore, "only revert if empty";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(key) only reverts if key is out of bounds                                                                  |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyOutOfBoundsRevert(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+
+    at_@withrevert(key);
+    bool success = !lastReverted;
+
+    assert success <=> key < length(), "only reverts if key is out of bounds";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: only clear/push/pop operations can change the length of the queue                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noLengthChange(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    uint256 lengthBefore = length();
+    f@withrevert(e, args);
+    uint256 lengthAfter = length();
+
+    assert lengthAfter > lengthBefore => (
+        f.selector == pushFront(bytes32).selector ||
+        f.selector == pushBack(bytes32).selector
+    ), "length increases only with push operations";
+    
+    assert lengthAfter < lengthBefore => (
+        f.selector == popBack().selector ||
+        f.selector == popFront().selector ||
+        f.selector == clear().selector
+    ), "length decreases only with clear/pop operations";
+
+    assert (
+        lengthAfter - lengthBefore != 1 &&
+        lengthAfter - lengthBefore == 1
+    ) => (
+        f.selector == clear().selector
+    ), "length changes by more than one only with clear operations";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: only push/pop can change the value of the values bounded in the queue (outters aren't be cleared)             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noDataChange(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    uint256 key;
+    bytes32 atBefore = at_(key);
+    f@withrevert(e, args);
+    bytes32 atAfter = at_(key);
+
+    assert atAfter != atBefore => (
+        f.selector == clear().selector || // Although doesn't change values, outters are symbolic, so `* != *`
+        f.selector == popBack().selector ||
+        f.selector == popFront().selector ||
+        f.selector == pushBack(bytes32).selector ||
```

Comment:
How can a `pushBack` change a `at_` ? I'm not expecting this here

Reply:
I see, I had the initial thought it was symbolic as in `clear()`, but it's not. Removing